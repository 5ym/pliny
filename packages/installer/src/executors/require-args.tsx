import { Box, Text, Newline, useApp } from 'ink'
import * as React from 'react'
import { EnterToContinue } from '../components/enter-to-continue'
import { useEnterToContinue } from '../utils/use-enter-to-continue'
import { useUserInput } from '../utils/use-user-input'
import { RecipeCLIArgs } from '../types'
import { Executor, executorArgument, ExecutorConfig, getExecutorArgument } from './executor'

export type RequiredArg = {
  name: string
  description: string
}

export interface Config extends ExecutorConfig {
  args: executorArgument<RequiredArg[]>
}

export const type = 'require-args'

const validateArgs = (requiredArgs: RequiredArg[], inputArgs: RecipeCLIArgs): RequiredArg[] => {
  const inputArgsName = new Set(Object.keys(inputArgs))
  return requiredArgs.filter((args) => {
    return !inputArgsName.has(args.name)
  })
}

export const Commit: Executor['Commit'] = ({ cliArgs, cliFlags, onChangeCommitted, step }) => {
  const userInput = useUserInput(cliFlags)
  const generatorArgs = React.useMemo(
    () => ({
      args: getExecutorArgument((step as Config).args, cliArgs),
      stepName: getExecutorArgument((step as Config).stepName, cliArgs),
    }),
    [cliArgs, step]
  )
  const [changeCommited, setChangeCommited] = React.useState(false)

  const handleChangeCommitted = React.useCallback(() => {
    setChangeCommited(true)
    onChangeCommitted(generatorArgs.stepName)
  }, [onChangeCommitted, generatorArgs])

  const childProps: CommitChildProps = {
    changeCommited,
    generatorArgs,
    handleChangeCommitted,
    cliArgs,
  }

  if (userInput) return <CommitWithInput {...childProps} />
  else return <CommitWithoutInput {...childProps} />
}

interface CommitChildProps {
  changeCommited: boolean
  generatorArgs: {
    args: RequiredArg[]
    stepName: string
  }
  handleChangeCommitted: () => void
  cliArgs: RecipeCLIArgs
}

const CommitWithInput = ({
  changeCommited,
  generatorArgs,
  handleChangeCommitted,
  cliArgs,
}: CommitChildProps) => {
  useEnterToContinue(handleChangeCommitted, !changeCommited)
  const { exit } = useApp()
  const unspecifiedArgs = validateArgs(generatorArgs.args, cliArgs)
  const requiredNames = generatorArgs.args.map((arg) => arg.name).join(', ')

  React.useEffect(() => {
    if (unspecifiedArgs.length !== 0) {
      const display = unspecifiedArgs.map((args) => `${args.name}: ${args.description}`).join('\n')
      exit(new Error(`The following required arguments are missing: \n${display}`))
    }
  })

  return (
    <Box flexDirection="column">
      <Text>Checking required arguments...</Text>
      {unspecifiedArgs.length === 0 && <Text>Using {requiredNames} in recipe...</Text>}
      <EnterToContinue />
    </Box>
  )
}

const CommitWithoutInput = ({
  changeCommited,
  generatorArgs,
  handleChangeCommitted,
  cliArgs,
}: CommitChildProps) => {
  const { exit } = useApp()
  const unspecifiedArgs = validateArgs(generatorArgs.args, cliArgs)
  const requiredNames = generatorArgs.args.map((arg) => arg.name).join(', ')

  React.useEffect(() => {
    if (!changeCommited) {
      handleChangeCommitted()
    }
  }, [changeCommited, handleChangeCommitted])

  React.useEffect(() => {
    if (unspecifiedArgs.length !== 0) {
      const display = unspecifiedArgs.map((args) => `${args.name}: ${args.description}`).join('\n')
      exit(new Error(`The following required arguments are missing: \n${display}`))
    }
  })

  return (
    <Box flexDirection="column">
      <Text>Checking required arguments...</Text>
      {unspecifiedArgs.length === 0 && <Text>Using {requiredNames} in recipe...</Text>}
    </Box>
  )
}
