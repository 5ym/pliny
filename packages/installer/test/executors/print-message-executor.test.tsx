/* eslint-disable @typescript-eslint/no-empty-function */
import { expect, describe, it } from 'vitest'
import { render } from 'ink-testing-library'
import React from 'react'
import stripAnsi from 'strip-ansi'
import { Commit as PrintMessageExecutor } from '../../src/executors/print-message-executor'

describe.skip('Executor', () => {
  const executorConfig = {
    stepId: 'printMessage',
    stepName: 'Print message',
    stepType: 'print-message',
    explanation: 'Testing text for a print message',
    message: 'My message',
  }
  it('should render PrintMessageExecutor', () => {
    const { lastFrame } = render(
      <PrintMessageExecutor
        cliArgs={{}}
        cliFlags={{ yesToAll: false }}
        onChangeCommitted={() => {}}
        step={executorConfig}
      />
    )

    expect(stripAnsi(lastFrame() as string)).toMatchSnapshot()
  })

  it('should contain a step name and explanation', () => {
    const { frames } = render(
      <PrintMessageExecutor
        cliArgs={{}}
        cliFlags={{ yesToAll: false }}
        onChangeCommitted={() => {}}
        step={executorConfig}
      />
    )

    expect(frames[0].includes('My message')).toBeTruthy()
  })
})
