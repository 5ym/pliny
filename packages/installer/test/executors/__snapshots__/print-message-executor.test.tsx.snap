// Vitest Snapshot v1

exports[`Executor > should render PrintMessageExecutor 1`] = `
"My message

Press ENTER to continue"
`;
